core/src/main/java/com/google/errorprone/refaster/UEnhancedForLoop.java

```diff
@@ -68,31 +128,12 @@ public Kind getKind() {
 
   @Override
   public JCEnhancedForLoop inline(Inliner inliner) throws CouldNotResolveImportException {
-    return makeForeachLoop(
-        inliner.maker(),
-        getVariable().inline(inliner),
-        getExpression().inline(inliner),
-        getStatement().inline(inliner));
-  }
-
-  private static JCEnhancedForLoop makeForeachLoop(
-      TreeMaker maker, JCVariableDecl variable, JCExpression expression, JCStatement statement) {
-    try {
-      if (RuntimeVersion.isAtLeast20()) {
-        return (JCEnhancedForLoop)
-            TreeMaker.class
-                .getMethod("ForeachLoop", JCTree.class, JCExpression.class, JCStatement.class)
-                .invoke(maker, variable, expression, statement);
-      } else {
-        return (JCEnhancedForLoop)
-            TreeMaker.class
-                .getMethod(
-                    "ForeachLoop", JCVariableDecl.class, JCExpression.class, JCStatement.class)
-                .invoke(maker, variable, expression, statement);
-      }
-    } catch (ReflectiveOperationException e) {
-      throw new LinkageError(e.getMessage(), e);
-    }
+    return inliner
+        .maker()
+        .ForeachLoop(
+            getVariable().inline(inliner),
+            getExpression().inline(inliner),
+            getStatement().inline(inliner));
```

Comment:
@cushon based on my tests (with JDK 20-ea build 28) the build also passes without these changes. Let me know if I should revert this part of the PR.

Reply:
I think I prepared https://github.com/google/error-prone/commit/df033f03cb55c3f26221660ac5ea49bbaeeab289 to fix an error in the internal CI, which was building at the Java 11 language level and APIs and then running on a newer JDK. We might not need this with your patch, though. Let me think about this a little harder.

(I think there was some related discussion in an external bug about how our CI is compiling and testing at each JDK version, but we're only releasing one set of artifacts built at a fixed version, and maybe we should have the CI do one build that matches the releases and then run the tests at different versions.)

Reply:
I see a regression when testing this compiled against the Java 11 APIs, and running on the latest version. I think that makes sense: the change to `TreeMaker#ForeachLoop` is source-compatible here, but if it resolves the Java 11 version at compile-time and sees a different version at runtime it will fail.

I think the reflection is still necessary, and I should also try to add that configuration to the external CI.

```
java.lang.NoSuchMethodError: 'com.sun.tools.javac.tree.JCTree$JCEnhancedForLoop com.sun.tools.javac.tree.TreeMaker.ForeachLoop(com.sun.tools.javac.tree.JCTree$JCVariableDecl, com.sun.tools.javac.tree.JCTree$JCExpression, com.sun.tools.javac.tree.JCTree$JCStatement)'
	at com.google.errorprone.refaster.UEnhancedForLoop.inline([UEnhancedForLoop.java:125](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/UEnhancedForLoop.java:125&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.UEnhancedForLoop.inline([UEnhancedForLoop.java:34](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/UEnhancedForLoop.java:34&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.USimpleStatement.inlineStatements([USimpleStatement.java:34](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/USimpleStatement.java:34&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.BlockTemplate.replace([BlockTemplate.java:215](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/BlockTemplate.java:215&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.BlockTemplate.replace([BlockTemplate.java:48](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/BlockTemplate.java:48&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.scan([RefasterScanner.java:126](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:126&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.scan([RefasterScanner.java:51](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:51&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce([TreeScanner.java:96](https://cs.corp.google.com/#search/&q=f:com/sun/source/util/TreeScanner.java:96&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.source.util.TreeScanner.visitMethod([TreeScanner.java:224](https://cs.corp.google.com/#search/&q=f:com/sun/source/util/TreeScanner.java:224&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.visitMethod([RefasterScanner.java:88](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:88&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.visitMethod([RefasterScanner.java:51](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:51&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept([JCTree.java:944](https://cs.corp.google.com/#search/&q=f:com/sun/tools/javac/tree/JCTree.java:944&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.visitClass([RefasterScanner.java:75](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:75&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.visitClass([RefasterScanner.java:51](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:51&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept([JCTree.java:851](https://cs.corp.google.com/#search/&q=f:com/sun/tools/javac/tree/JCTree.java:851&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.source.util.TreeScanner.scan([TreeScanner.java:92](https://cs.corp.google.com/#search/&q=f:com/sun/source/util/TreeScanner.java:92&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.scan([RefasterScanner.java:132](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:132&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.scan([RefasterScanner.java:51](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:51&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.source.util.TreeScanner.scan([TreeScanner.java:111](https://cs.corp.google.com/#search/&q=f:com/sun/source/util/TreeScanner.java:111&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce([TreeScanner.java:119](https://cs.corp.google.com/#search/&q=f:com/sun/source/util/TreeScanner.java:119&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.source.util.TreeScanner.visitCompilationUnit([TreeScanner.java:152](https://cs.corp.google.com/#search/&q=f:com/sun/source/util/TreeScanner.java:152&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCCompilationUnit.accept([JCTree.java:619](https://cs.corp.google.com/#search/&q=f:com/sun/tools/javac/tree/JCTree.java:619&ws=cushon/225130&snapshot=5))
	at jdk.compiler/com.sun.source.util.TreeScanner.scan([TreeScanner.java:92](https://cs.corp.google.com/#search/&q=f:com/sun/source/util/TreeScanner.java:92&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterScanner.scan([RefasterScanner.java:132](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterScanner.java:132&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.RefasterRule.apply([RefasterRule.java:139](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/RefasterRule.java:139&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.CodeTransformerTestHelper.transform([CodeTransformerTestHelper.java:75](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/CodeTransformerTestHelper.java:75&ws=cushon/225130&snapshot=5))
	at com.google.errorprone.refaster.TemplateIntegrationTest.expectTransforms([TemplateIntegrationTest.java:68](https://cs.corp.google.com/#search/&q=f:com/google/errorprone/refaster/TemplateIntegrationTest.java:68&ws=cushon/225130&snapshot=5))
```

Reply:
I could not reproduce this locally, but the stack trace speaks for itself. I added a commit to revert this section :+1: